core/include/core/index_geometry.hpp

```diff
@@ -0,0 +1,446 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+#include <iterator>
+#include <sstream>
+#include <string>
+#include <utility>
+
+#include "core/surface_base.hpp"
+#include "masks/masks.hpp"
+#include "utils/enumerate.hpp"
+#include "utils/indexing.hpp"
+
+namespace detray {
+/**
+ * @brief Index geometry implementation
+ *
+ * This class provides a geometry that defines logic volumes which contain
+ * the detector surfaces, joined together by dedicated portal surfaces. It
+ * exports all types needed for navigation and strictly only keeps the
+ * index data (links) that define the geometry relations.
+ *
+ * @tparam array_type the type of the internal array, must have STL
+ *                    semantics
+ * @tparam vector_type the type of the internal array, must have STL
+ *                     semantics
+ * @tparam surface_source_link the type of the link to an external surface
+ *                             source
+ *
+ * @note The geometry knows nothing about coordinate systems. This is
+ *       handeled by geometry access objects (e.g. the grid).
+ */
+template <template <typename, unsigned int> class array_type = darray,
+          template <typename> class vector_type = dvector,
+          template <typename...> class tuple_type = dtuple,
+          typename surface_source_link = dindex,
+          typename bounds_source_link = dindex>
+class index_geometry {
+
+    public:
+    // Known primitives
+    enum objects : bool {
+        e_surface = true,
+        e_portal = false,
+    };
+
+    /** Encodes the position in a collection container for the respective
+        mask type . */
+    enum mask_container_index : unsigned int {
+        e_mask_types = 5,
+        e_rectangle2 = 0,
+        e_trapezoid2 = 1,
+        e_annulus2 = 2,
+        e_cylinder3 = 3,
+        e_ring2 = 4,
+        e_single3 = std::numeric_limits<unsigned int>::max(),
+        e_unknown = std::numeric_limits<unsigned int>::max(),
+    };
+
+    /// Portals components:
+    /// - links:  next volume, next (local) object finder
+    using portal_links = array_type<dindex, 2>;
+    /// - masks, with mask identifiers 0, 1
+    using portal_cylinder = cylinder3<false, cylinder_intersector,
+                                      __plugin::cylindrical2, portal_links, 0>;
+    using portal_disc =
+        ring2<planar_intersector, __plugin::cartesian2, portal_links, 1>;
+    // - mask index: type, { first/last }
+    using portal_mask_index = tuple_type<dindex, array_type<dindex, 2>>;
+    using portal_mask_container =
```

Comment:
For future, we can actually make the portal container contain all masks at some stage, this is now the case for the cylinder volumes.

Reply:
One way to do this is #103, but then we confuse the navigator :)

Reply:
Do you something like this: https://github.com/acts-project/detray/pull/109/files#diff-1922c13a0370f0927decfa996bcb57b35996e1b30a1d61846608bb14b839721c
which contains all amsk types?